### ISSUE TITLE
Logging error on Hydra multiruns using Pytorch lightning

### ISSUE THREAD

**Issue by tadej-redstone** [wandb/client#1314]
When I do a multirun on Hydra (with my model using Pytorch lightning), the program always crashes after (at the end of) the second run.
The problem seems to be Wandb closing a logger prematurely - if I remove Wandb Logger the problem dissapears.

I don't have a simple reproducible example (however I can provide more info if necessary), but my main file is very simple:
```python
from omegaconf import DictConfig, OmegaConf, open_dict
import hydra

from classification.data import Data
from classification.model import Classifier
from classification.utils import get_trainer

@hydra.main(config_name="config")
def run_train(hparams: DictConfig):
    
    data = Data(batch_size=hparams.batch_size)

    data.prepare_data()
    data.setup("fit")
    
    with open_dict(hparams):
        hparams.len_train = len(data.train_dataloader())

    model = Classifier(hparams)
    
    trainer = get_trainer(hparams)    
    
    trainer.fit(model, data)


if __name__ == "__main__":
    run_train()
```

Here's the error log:
<details>
  <summary>Error log</summary>
  
  ```
  --- Logging error ---
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/logging/__init__.py", line 1084, in emit
    stream.write(msg + self.terminator)
  File "/opt/conda/lib/python3.8/site-packages/wandb/lib/redirect.py", line 23, in write
    self.stream.write(data)
ValueError: I/O operation on closed file.
Call stack:
  File "hydra_run.py", line 31, in <module>
    run_train()
  File "/opt/conda/lib/python3.8/site-packages/hydra/main.py", line 32, in decorated_main
    _run_hydra(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 354, in _run_hydra
    run_and_report(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 198, in run_and_report
    return func()
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/utils.py", line 355, in <lambda>
    lambda: hydra.multirun(
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/hydra.py", line 136, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/core_plugins/basic_sweeper.py", line 154, in sweep
    results = self.launcher.launch(batch, initial_job_idx=initial_job_idx)
  File "/opt/conda/lib/python3.8/site-packages/hydra/_internal/core_plugins/basic_launcher.py", line 76, in launch
    ret = run_job(
  File "/opt/conda/lib/python3.8/site-packages/hydra/core/utils.py", line 125, in run_job
    ret.return_value = task_function(task_cfg)
  File "hydra_run.py", line 24, in run_train
    trainer.fit(model, data)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/states.py", line 48, in wrapped_fn
    result = fn(self, *args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1073, in fit
    results = self.accelerator_backend.train(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/accelerators/gpu_backend.py", line 51, in train
    results = self.trainer.run_pretrain_routine(model)
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1239, in run_pretrain_routine
    self.train()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 416, in train
    self.run_training_teardown()
  File "/opt/conda/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 1136, in run_training_teardown
    log.info('Saving latest checkpoint..')
Message: 'Saving latest checkpoint..'
Arguments: ()
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/hydra/core/utils.py", line 247, in _flush_loggers
    h_weak_ref().flush()
  File "/opt/conda/lib/python3.8/logging/__init__.py", line 1065, in flush
    self.stream.flush()
ValueError: I/O operation on closed file.

  ```
</details>

**Comment by tyomhak** [comment]
Hi @tadej-redstone , thanks for flagging this issue. We're looking into this.

**Comment by omry** [comment]
Hi guys,
I can add a try block around the flush there to suppress the error but I would like to see a minimal repro first.
Anything unusual going on with the logging in PL or the wandb logger than can shed some light on what's happening here?

**Comment by tadejsv** [comment]
Hey @omry, here's a minimal repository that reproduces the error:
[https://github.com/tadejsv/wandb-log-bug](https://github.com/tadejsv/wandb-log-bug)

**Comment by omry** [comment]
@tadejsv, thanks.
By minimal repro I meant something without external dependencies.

looking at the [PL Wandb logger](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/wandb.py) I am not seeing anything related to logging.

Can you try to repro without PL involvement? (using the wandb API directly).

**Comment by omry** [comment]
I filed [this](https://github.com/facebookresearch/hydra/issues/1053), but it would be great if you can help isolate the root cause.

**Comment by tadejsv** [comment]
Thanks for taking a look at it. I've already answered under hydra issue, but I'll do it here too since the bug is probably on wandb's side:

I've [isolated](https://github.com/tadejsv/wandb-log-bug) the error, and it seems to be caused by the `reinit=True` option in `wandb.init()` (this is what PL does by default). Setting it to `False` removes the error.

**Comment by omry** [comment]
Great. From the Hydra side I will turn exceptions during flushing of the loggers into a warning, but it does suggest an underlying problem in which might cause logging issues in the next run when doing multirun with the basic launcher.

**Comment by borisdayma** [comment]
`reinit` is used to ensure we start a new W&B run. Otherwise it could be a noop when there is already an active W&B run (which typically closes automatically at the end of the script). In those cases, `reinit` will first close the active run (if present) and then create a new one.

I tried manually calling `wandb.finish()` at the end of the run and it solves the issue. Here is a [minimal example](https://gist.github.com/borisdayma/3e54987c55da0ca54f3142e3ed4e8a91) based on the one from @tadejsv 
I don't think we should automatically call `wandb.finish()` after a pytorch-lightning training loop as users could manually want to log other items (summary & interpretation metrics for example).

I think hydra [tries to flash all loggers](https://github.com/facebookresearch/hydra/blob/78fc70ecc746fc70926d27737bdd7e34ca7bff48/hydra/core/utils.py#L245) while there is typically one for `stdout` and 2 managed by `wandb`which could be the cause of this issue.

@omry If the intent is just to flush stdout, could we just do it on `logging._handlerList[0]`?

**Comment by omry** [comment]
The intent is to flush all loggers, not stdout.

**Comment by tadej-redstone** [comment]
@borisdayma I think PL should call `wandb.finish()` after the end of training. It's either that (with the user performing end of run logging in `on_train_end` callback method), 
or
documenting that the user has to manually call `wandb.finish()` (which is not even the "correct" way of doing it, since the user would need to extract the wandb run from the `Trainer` object) in his script.

**Comment by omry** [comment]
@borisdayma, can you explain the interaction the wandb client has with python logging?
1. What is it achieving?
2. When does it initialize? when does it un-initialize?

The use case with Hydra is that two different jobs are running in serial in the same process (There are launcher plugins that allows them to run in parallel in different processes, but the default is serial in the same process).
The motivation for flushing the loggers is that in the past I have seen cases where the logger remains unflushed for a long time and the user cannot see the complete output of the first job while the second is already running.

Somehow, when using wandb in this scenario (which I have not yet looked at carefully) - the log file for wandb is closed, but the logger remains in the Python logging facade, which leads to an error when Hydra is trying to flush it.

**Comment by borisdayma** [comment]
Thanks @omry for the clarifications. It makes sense.
We're looking into a workaround that would not require closing the run. cc @raubitsj

**Comment by tomsercu** [comment]
Hi folks, I'm still  seeing this issue if you do more than 2 runs with multirun. See [this gist](https://gist.github.com/tomsercu/72acb8a435afcda0043ba351536d821a ), where i adapted @borisdayma 's example to make it a bit more verbose. 
System: Python 3.7.9, wandb 0.10.10, hydra 1.0.3.

You can see the output is printed in the first 2 runs but disappears for the 3d (and all consecutive) runs. 
Output disappears both for plain `print` statements and using the hydra-configured python logger.
And a similar error is raised `OSError: [Errno 9] Bad file descriptor`.
In real runs, it seems that the jobs still run and sync to wandb even when all stdout is gone.

**Comment by qshzh** [comment]
@tomsercu  I encountered the same issue. Do you figure it out or maybe we can file a new issue?

**Comment by tomsercu** [comment]
Nope, I have no progress on this issue.
However, the issue doesn't occur when running with hydra's `--multirun` and the **slurm launcher** (note default launcher is running things locally sequentially in the same process). Quickly looking at the `hydra/launcher=joblib`, I think this would probably avoid the issue too -- since Joblib.parallel spins up several python subprocesses.

**Comment by lucmos** [comment]
> Nope, I have no progress on this issue.
> However, the issue doesn't occur when running with hydra's `--multirun` and the **slurm launcher** (note default launcher is running things locally sequentially in the same process). Quickly looking at the `hydra/launcher=joblib`, I think this would probably avoid the issue too -- since Joblib.parallel spins up several python subprocesses.

Unfortunately, I'm encountering the same problem, and the joblib launcher does not avoid the issue.

**Comment by vanpelt** [comment]
Hey @lucmos with the most recent release of wandb, can you try setting the `WANDB_START_METHOD` environment variable to "thread" and see if you're still having the same issue?

We're also doing some work on multiprocessing and calling wandb.init in both a parent and child process that could be relevant here.  If you could provide us with a colab or code snippet to reproduce the issue it would really help us get to the bottom of it.

**Comment by lucmos** [comment]
Hello @vanpelt,

Good news, with `WANDB_START_METHOD=thread` the minimum working example I prepared works.
(in case of further problems on production code that uses lightning I'll post again here)

You can find the mwe [here](https://gist.github.com/lucmos/534c14a3ed7fd6ed651962a41680ccc7)

Are there any contraindications in using always `WANDB_START_METHOD=thread`?

**Comment by vanpelt** [comment]
Hey @lucmos we just added this start method so we're still evaluating performance around it.  The default start method is spawn which has better semantics / logic seperation but can interact poorly especially if your code is leveraging multiprocessing as well.  We're planning to atleast detect this case and automatically use `thread` when appropriate.

**Comment by ariG23498** [comment]
Hey folks

Closing this issue due to inactivity.
With further issues pertaining to this ticket, please feel free to ask to reopen this thread. Thanks!

**Comment by andreabac3** [comment]
@lucmos Thank you, `WANDB_START_METHOD=thread` solve my issue.

**Comment by simpleParadox** [comment]
I'm using 
```
lightning=2.40
wandb=0.18.7
hydra=1.3.2
```

I found that the when running hydra -multirun sweeps (using the inbuilt sweeper - that starts runs locally and sequentially), using os.environ["WANDB_START_METHOD"]="thread" does not start a new wandb run for different hyperpamater configurations.

I had to manually call `wandb.finish()` inside the train function (but not after the train function is called) to make wandb start a new session.

Example:
```
import wandb
from lightning.pytorch.loggers import WandbLogger
from lightning.pytorch import Trainer


@hydra.main(version_base=None, config_path="configs", config_name="config") # This is the path to the 'config' file.
def train():
    wandb_logger = WandbLogger(name=cfg.dataset_config.run_name,
        # log_model="all", 
        config=omegaconf.OmegaConf.to_container(
            cfg, resolve=True, throw_on_missing=True
        ),
        project="simpleparadox_limitations_of_transformers",
        entity="simpleparadox",
    )
   # Do training stuff here.
   # prep data
   trainer = Trainer()
   trainer.fit()

   # Call wandb finish.
   wandb.finish()

if __name__ == "__main__":
    train()

    # Do not call wandb.finish() here but at the end of the train() function.

```
